dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/mapper/ProgramStageInstanceMapper.java

```diff
@@ -65,7 +68,40 @@ public ProgramStageInstance map( Event event )
     {
         ProgramStageInstance psi = workContext.getProgramStageInstanceMap().get( event.getUid() );
 
-        return psi == null ? mapForInsert( event ) : mapForUpdate( event, psi );
+        if ( psi == null )
+        {
+            psi = mapForInsert( event );
+        }
+        else
+        {
+            mapForUpdate( event, psi );
+        }
+
+        return mapDataValueIdentifier( psi, super.workContext.getDataElementMap() );
+    }
+
+    /**
+     * Set data element identifier to uid. We run the psi through this method to
+     * make sure data value identifiers are UID. Otherwise, we would store data
+     * values with whatever identifier is used for import, for example CODE or
+     * ATTRIBUTE.
+     *
+     * @param psi the psi with the datavalues to normalize to uid identifiers
+     * @param dataElementMap a map of persisted data elements.
+     * @return a program stage instance with normalized data values.
+     */
+    private ProgramStageInstance mapDataValueIdentifier( ProgramStageInstance psi,
+        Map<String, DataElement> dataElementMap )
+    {
+        for ( EventDataValue dv : psi.getEventDataValues() )
+        {
+            DataElement de = dataElementMap.get( dv.getDataElement() );
+            if ( de != null )
+            {
+                dv.setDataElement( de.getUid() );
```

Comment:
should we do something else in case there is a `dv.getDataElement()` (non-null) but it was not found in the `dataElementMap`?

Reply:
If de is null, it means it did not exists in the database, and it should already be caught by the validation, so if we got to this code with that case, we would have to ignore it anyway, we can't really recover from it in that case.

I'll have a look about making a small test for this